Adding a blog post for AWS 5.0.0 release and the new S3 Bucket work (#1256)
pulumi/pulumi-hugo: themes/default/content/blog/announcing-v5.0.0-of-the-pulumi-aws-provider/index.md

---
title: "Announcing v5.0.0 of the Pulumi AWS Provider"
date: 2022-03-23
meta_desc: "v5.0.0 of the Pulumi AWS Provider is now available"
meta_image: meta.png
authors:
- paul-stack
tags:
- aws
---

> We are excited to announce v5.0.0 of the Pulumi AWS Provider. Pulumi prides itself on how it continually tries to ensure
> that major version releases are compatible with existing applications. The upstream provider that this is based upon, has had
> a major breaking change in the S3 Bucket resource. In order to ensure that our users are not impacted, We have extended
> our provider to provider both the old implementation and the new implementation of the S3 Bucket resource.
> This means that all existing usages of S3 Bucket will continue to work as normal. We also now offer the ability to
> create a v2 variants of the S3 Bucket.

Reviewer: ```suggestion
We are excited to announce v5.0.0 of the Pulumi AWS Provider. Pulumi prides itself on ensuring
that major version releases are compatible with existing applications. The upstream provider this is based upon had
a major breaking change in the S3 Bucket resource. We have extended
our provider to include both the old and new implementations of the S3 Bucket resource to prevent users from being impacted.
All existing S3 Buckets will continue to work. We also now offer the ability to
create a v2 variant of the S3 Bucket.
```

    },
});
```

{{% /choosable %}}

{{% choosable language python %}}

```python
from pulumi_aws import s3

web_bucket = s3.Bucket('s3-website-bucket',
                       website=s3.BucketWebsiteArgs(
                       index_document="index.html",
                       ))
```

{{% /choosable %}}

{{% choosable language csharp %}}

```csharp
using Pulumi;
using Pulumi.Aws.S3;
using Pulumi.Aws.S3.Inputs;

class WebsiteStack : Stack
{
    public WebsiteStack()
    {
        var bucket = new Bucket("my-bucket", new BucketArgs
        {
            Website = new BucketWebsiteArgs
            {
                IndexDocument = "index.html"
            }
        });
    }
}
```

{{% /choosable %}}

{{% choosable language go %}}

```go
package main

import (
	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/s3"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		_, err := s3.NewBucket(ctx, "s3-website-bucket", &s3.BucketArgs{
			Website: s3.BucketWebsiteArgs{
				IndexDocument: pulumi.String("index.html"),
			},
		})
		if err != nil {
			return err
		}
		
		return nil
	})
}
```

{{% /choosable %}}

{{% /chooser %}}

> In the upstream provider, the S3 Bucket resource has been broken into multiple pieces meaning that you would need to move the website
> part of the bucket out to be it's own resource. *We have ensured that this will continue to work in Pulumi.*

Reviewer: ```suggestion
The S3 Bucket resource has been split into multiple pieces in the upstream provider, which means you would need to move the website
part of the bucket out to be its own resource. *We have ensured that this will continue to work in Pulumi.*
```

	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/s3"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		siteBucket, err := s3.NewBucket(ctx, "s3-website-bucket", nil)
		if err != nil {
			return err
		}

		_, err := s3.NewBucketWebsiteConfigurationV2(ctx, "website-config", &s3.BucketWebsiteConfigurationV2Args{
			Bucket: siteBucket.bucket,
			IndexDocument: &s3.BucketWebsiteConfigurationV2IndexDocumentArgs{
				Suffix: pulumi.String("index.html"),
			},
		})
		if err != nil {
			return err
        }
		
		return nil
	})
}
```

{{% /choosable %}}

{{% /chooser %}}

## Can I migrate from `aws.s3.Bucket` to `aws.s3.BucketV2`?

Yes! As part of this major version, we have taken advantage of using [aliases]({{< relref "/docs/intro/concepts/resources/options/aliases" >}}) in Pulumi to be able to allow users to
change to `aws.s3.BucketV2` easily. An alias was added to the new `aws.s3.BucketV2` resource as follows:

```go
    "aws_s3_bucket": {
        Tok: awsResource(s3Mod, "BucketV2"),
        Aliases: []tfbridge.AliasInfo{
            {
                Type: stringRef("aws:s3/bucket:Bucket"),
            },
        },
    },
```

> This ensures that when a user changes from `aws.s3.Bucket` to `aws.s3.BucketV2` that the Pulumi engine understands that
> the bucket is not a new bucket but is an existing bucket that has changed. Therefore, the Pulumi engine will not destroy
> the bucket and recreate it. This means that to change from 1 codebase to the other, we just need to keep the names of the
> resources the same (e.g `s3-website-bucket` in the code above) and then continue to add the new code to satisfy the way the
> resources need to be handed (e.g. `BucketWebsiteConfigurationV2` in the code above).

Reviewer: ```suggestion
When a user changes from `aws.s3.Bucket` to `aws.s3.BucketV2`, the Pulumi engine understands that
the bucket is not new but is an existing bucket that has changed. The Pulumi engine will not destroy
the bucket and recreate it. To change from 1 codebase to the other, we need to keep the names of the
resources the same (e.g., `s3-website-bucket` in the code above) and then continue to add the new code to satisfy the way the
resources need to be handled (e.g. `BucketWebsiteConfigurationV2` in the code above).
```

You can see the steps on how this works as follows:

{{< asciicast id="YMgxBRkaBv95iegpkgMjBy7Zt" >}}

> Notice that the bucket is not recreate - the only change is the addition of the new aws s3 resource for the website configuration.
> Your existing infrastructure is not going to change and thus we can easily migrate between the different resource types.

Reviewer: ```suggestion
The bucket is not recreated - the only change is the new aws s3 resource for the website configuration.
Your existing infrastructure is not going to change and we can easily migrate between the different resource types.
```
